internal/indexer/walker.go

```diff
@@ -111,44 +110,39 @@ func (idx *Indexer) WalkedModule(ctx context.Context, modHandle document.DirHand
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		} else {
-			ids = append(ids, id)
-			blockingJobIds = append(blockingJobIds, id)
+			ids = append(ids, pSchemaId)
+			refCollectionDeps = append(refCollectionDeps, pSchemaId)
 		}
 	}
 
 	if dataDir.ModuleManifestPath != "" {
 		// References are collected *after* manifest parsing
 		// so that we reflect any references to submodules.
-		id, err := idx.jobStore.EnqueueJob(job.Job{
+		modManifestId, err := idx.jobStore.EnqueueJob(job.Job{
 			Dir: modHandle,
 			Func: func(ctx context.Context) error {
 				return module.ParseModuleManifest(idx.fs, idx.modStore, modHandle.Path())
 			},
-			Type:  op.OpTypeParseModuleManifest.String(),
-			Defer: idx.decodeInstalledModuleCalls(modHandle),
+			Type: op.OpTypeParseModuleManifest.String(),
+			Defer: func(ctx context.Context, jobErr error) (job.IDs, error) {
+				return idx.decodeInstalledModuleCalls(modHandle)
+			},
 		})
 		if err != nil {
 			errs = multierror.Append(errs, err)
 		} else {
-			ids = append(ids, id)
-			blockingJobIds = append(blockingJobIds, id)
+			ids = append(ids, modManifestId)
+			refCollectionDeps = append(refCollectionDeps, modManifestId)
 		}
 	}
 
-	// Here we wait for all dependent jobs to be processed to
-	// reflect any data required to collect reference origins.
-	// This assumes scheduler is running to consume the jobs
-	// by the time we reach this point.
-	err = idx.jobStore.WaitForJobs(ctx, blockingJobIds...)
-	if err != nil {
-		return ids, err
-	}
-
-	rIds, err := idx.collectReferences(ctx, modHandle)
-	if err != nil {
-		errs = multierror.Append(errs, err)
-	} else {
-		ids = append(ids, rIds...)
+	if parseId != "" {
```

Comment:
Do we need module metadata before collecting references? Should we check for `&& metaId != ""` here?

Reply:
Same answer as above - strictly speaking we don't, but it's useful if we have them. This logic basically applies progressive enhancement principle in that we still collect references, even if scheduling of metadata has failed, there will just likely be less references collected as a result.